Fix GKE Helm options for CI and docs.

jrajahalme opened the pull request:
GKE install guide specifies the "ipam.config=kubernetes" option while
CI gkeHelmOverrides do not. Adding this option to CI gkeHelmOverrides
allowed Ginkgo runs in GKE to succeed.

Helm name for the native routing CIDR is
"global.nativeRoutingCIDR". Cilium agent command line option name is
"native-routing-cidr".

Fixes: #12053

jrajahalme committed:
docs: Fix GKE install native routing CIDR option.

Helm name for the native routing CIDR is
"global.nativeRoutingCIDR". Cilium agent command line option name is
"native-routing-cidr".

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
ci: Add ipam.config=kubernetes for gkeHelmOverrides

GKE install guide specifies the "ipam.config=kubernetes" option while
CI gkeHelmOverrides do not. Adding this option to CI gkeHelmOverrides
allowed Ginkgo runs in GKE to succeed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on Documentation/gettingstarted/k8s-install-gke.rst:
```diff
@@ -92,7 +92,7 @@ below. This will ensure all pods are managed by Cilium.
       --set global.gke.enabled=true \\
       --set config.ipam=kubernetes \\
-      --set global.native-routing-cidr=$NATIVE_CIDR
+      --set global.nativeRoutingCIDR=$NATIVE_CIDR
```
Note that we need this particular change on v1.7 as well (but not the ipam option). Given you're on backports this week, probably easiest for you to make sure the right commit from this PR goes back.

joestringer replied:
Actually, it looks like we backported this change to the docs but never backported the helm side interpretation so it has been a no-op the entire time :facepalm:

jrajahalme committed:
test: Use generic operator image in GKE

Using the "fat" cilium-operator image fails with:

  State:          Waiting
    Reason:       CrashLoopBackOff
  Last State:     Terminated
    Reason:       ContainerCannotRun
    Message:      OCI runtime create failed: container_linux.go:345: starting container process caused "exec: \"cilium-operator-generic\": executable file not found in $PATH": unknown
    Exit Code:    127

Looks like the "cilium-operator" image tries to exec "cilium-operator-generic"?

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Get detail on failing kube-dns

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Pass Hubble-relay image to ginkgo on GKE

Pass the full image name to Ginkgo in the GKE Jenkinsfile, as
otherwise Ginkgo tries to use the default tag (latest), which is not
available for the test setup.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Make GKE delete namespaces script more portable and note it in docs

Basic regular expressions in sed do not support '+', while the command
line arguments for extended regular expressions are different for GNU
sed and Mac OSX. Use (unquoted) '*' instead, as there is no harm
replacing '[]' with '[]'. This form work both on Linux and Mac OSX.

Add a note to the e2e testing docs so that users know to run the
script if namespaces get stuck at 'Terminating'.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -140,6 +140,7 @@ var (
 		"global.nodePort.mode":        "snat",
 		"global.gke.enabled":          "true",
 		"global.nativeRoutingCIDR":    "10.0.0.0/8",
+		"config.ipam":                 "kubernetes",
```
FWIW in the current configmap, the `.Values.global.gke.enabled` also configures this. It's defined earlier in the file as well based on this `config.ipam` setting but I'm not sure about the precedence between the multiple declarations.

jrajahalme replied:
Good point, I'll remove this and re-test now that I know how to deal with the namespaces stuck at "Terminating" state!

jrajahalme replied:
Tested, does not work without `config.ipam: "kubernetes"`. Here's the diff on the CI generated cilium yaml, from with to without:
```
$ diff cilium-1619663866a24408.yaml cilium-1619670fe7897668.yaml
237a238
>   k8s-require-ipv6-pod-cidr: "false"
256c257,259
<   ipam: "kubernetes"
---
>   ipam: "cluster-pool"
>   cluster-pool-ipv4-cidr: "10.0.0.0/8"
>   cluster-pool-ipv4-mask-size: "24"
```

jrajahalme committed:
CI: Create GKE cluster lock in namespace 'cilium-ci-lock'

Create the cluster lock in namespace 'cilium-ci-lock' instead of
'default' so that it survives the CI deleting all resources in the
default namespace:

15:57:58  22:57:57 STEP: Deleting rs [lock-6f4fb6cdfb] in namespace default
15:57:58  22:57:57 STEP: Waiting for 1 deletes to return (lock-6f4fb6cdfb)
15:57:58  22:57:57 STEP: Deleting deployment [lock] in namespace default
15:57:58  22:57:57 STEP: Waiting for 1 deletes to return (lock)

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Increase Apply timeout

GKE can time out on apply when installing Cilium, use a longer timeout.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Do not set 'global.devices' for GKE

Without this override we'd get the following in the generated Cilium yaml:

$ diff -C3 good-cilium.yaml cilium-1619730510c46898.yaml
*** good-cilium.yaml	2020-06-17 14:07:44.000000000 -0700
--- cilium-1619730510c46898.yaml	2020-06-17 14:46:51.000000000 -0700
***************
*** 224,229 ****
--- 224,232 ----
    install-iptables-rules: "true"
    auto-direct-node-routes: "false"
    native-routing-cidr: 10.0.0.0/8
+   # List of devices used to attach bpf_host.o (implements BPF NodePort,
+   # host-firewall and BPF masquerading)
+   devices: "eth0 eth0\neth0"
    kube-proxy-replacement:  "probe"
    node-port-mode: "snat"
    node-port-bind-protection: "true"

Fixes: #11969
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -143,6 +143,8 @@ var (
 		"global.nativeRoutingCIDR":    "10.0.0.0/8",
 		"global.hostFirewall":         "false",
+		"config.ipam":                 "kubernetes",
+		"global.devices":              "", // Override "eth0 eth0\neth0"
```
But where does the `eth0 eth0\neth0` come from? Presumably somewhere else in the CI code, right?

jrajahalme replied:
#11969

jrajahalme replied:
My guess is that the effect of setting 'global.devices' is not tested on the GKE datapath configuration.

joestringer replied:
Looks like it yeah. I'm OK with this workaround for now, but /cc @pchaigno for figuring out how we want to tackle this in v1.9.

pchaigno replied:
The incorrect `global.devices` will be fixed by #12176. We should probably revert the devices overwrite once it's merged.

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -1401,7 +1403,7 @@ func (kub *Kubectl) Apply(options ApplyOptions) *CmdRes {
 		cmd = options.Piped + " | " + cmd
 	}
-	return kub.ExecMiddle(cmd)
+	return kub.Exec(cmd)
```
What's the timeout bump here? 30s->4m? This also extends the time to determine a legitimate failure when there's clearly a failure.

jrajahalme replied:
Maybe an `ExecLongish()` with a 1 min timeout would be better?

jrajahalme replied:
Would this be cleaner:
```
	ctx, cancel := context.WithTimeout(context.Background(), MidCommandTimeout * 2)
	defer cancel()
	return kub.ExecContext(ctx, cmd)
```
